File: apps/admin-panel/lib/constants/terms.ts
import { InterestInterval, Period, DisbursalPolicy } from "../graphql/generated"

export const TERMS_FIELD_LIMITS = {
  annualRate: { min: 0, max: 50 },
  durationUnits: { min: 1, max: 120 },
  oneTimeFeeRate: { min: 0, max: 10 },
  initialCvl: { min: 100, max: 500 },
  marginCallCvl: { min: 100, max: 500 },
  liquidationCvl: { min: 100, max: 500 },
} as const

export type TermsFieldsToValidate = {
  annualRate: string
  durationUnits: string
  oneTimeFeeRate: string
  initialCvl: string
  marginCallCvl: string
  liquidationCvl: string
}

export function validateTermsFields(values: TermsFieldsToValidate): string | null {
  const annualRate = parseFloat(values.annualRate)
  const durationUnits = parseFloat(values.durationUnits)
  const oneTimeFeeRate = parseFloat(values.oneTimeFeeRate)
  const initialCvl = parseFloat(values.initialCvl)
  const marginCallCvl = parseFloat(values.marginCallCvl)
  const liquidationCvl = parseFloat(values.liquidationCvl)

  if (isNaN(annualRate) || annualRate < TERMS_FIELD_LIMITS.annualRate.min || annualRate > TERMS_FIELD_LIMITS.annualRate.max)
    return "annualRateRange"
  if (isNaN(durationUnits) || durationUnits < TERMS_FIELD_LIMITS.durationUnits.min || durationUnits > TERMS_FIELD_LIMITS.durationUnits.max)
    return "durationRange"
  if (isNaN(oneTimeFeeRate) || oneTimeFeeRate < TERMS_FIELD_LIMITS.oneTimeFeeRate.min || oneTimeFeeRate > TERMS_FIELD_LIMITS.oneTimeFeeRate.max)
    return "oneTimeFeeRateRange"
  if (isNaN(initialCvl) || initialCvl < TERMS_FIELD_LIMITS.initialCvl.min || initialCvl > TERMS_FIELD_LIMITS.initialCvl.max)
    return "initialCvlRange"
  if (isNaN(marginCallCvl) || marginCallCvl < TERMS_FIELD_LIMITS.marginCallCvl.min || marginCallCvl > TERMS_FIELD_LIMITS.marginCallCvl.max)
    return "marginCallCvlRange"
  if (isNaN(liquidationCvl) || liquidationCvl < TERMS_FIELD_LIMITS.liquidationCvl.min || liquidationCvl > TERMS_FIELD_LIMITS.liquidationCvl.max)
    return "liquidationCvlRange"
  if (initialCvl <= marginCallCvl || marginCallCvl <= liquidationCvl)
    return "cvlOrderError"

  return null
}

export const DEFAULT_TERMS = {
  OBLIGATION_LIQUIDATION_DURATION_FROM_DUE: {
    UNITS: 60,
    PERIOD: Period.Days,
  },
  OBLIGATION_OVERDUE_DURATION_FROM_DUE: {
    UNITS: 50,
    PERIOD: Period.Days,
  },
  INTEREST_DUE_DURATION_FROM_ACCRUAL: {
    UNITS: 0,
    PERIOD: Period.Days,
  },
  ACCRUAL_CYCLE_INTERVAL: InterestInterval.EndOfMonth,
  ACCRUAL_INTERVAL: InterestInterval.EndOfDay,
  DURATION_PERIOD: Period.Months,
  DISBURSAL_POLICY: DisbursalPolicy.SingleDisbursal,
}

File: lana/admin-graphql-credit/src/terms.rs
use async_graphql::*;
use serde::{Deserialize, Serialize};

pub use admin_graphql_shared::credit::{
    CVLPct, Duration, FiniteCVLPct, InfiniteCVLPct, Period, TermValues,
};

pub use lana_app::terms::{
    AnnualRatePct, CVLPct as DomainCVLPct, DisbursalPolicy, FacilityDuration as DomainDuration,
    InterestInterval, ObligationDuration as DomainObligationDuration, OneTimeFeeRatePct,
    TermValues as DomainTermValues,
};

#[derive(InputObject)]
pub struct TermsInput {
    pub annual_rate: AnnualRatePct,
    pub accrual_interval: InterestInterval,
    pub accrual_cycle_interval: InterestInterval,
    pub one_time_fee_rate: OneTimeFeeRatePct,
    pub disbursal_policy: DisbursalPolicy,
    pub duration: DurationInput,
    pub interest_due_duration_from_accrual: DurationInput,
    pub obligation_overdue_duration_from_due: DurationInput,
    pub obligation_liquidation_duration_from_due: DurationInput,
    pub margin_call_cvl: CVLPctValue,
    pub initial_cvl: CVLPctValue,
    pub liquidation_cvl: CVLPctValue,
}

#[derive(InputObject)]
pub struct DurationInput {
    pub period: Period,
    pub units: u32,
}

impl From<DurationInput> for DomainDuration {
    fn from(duration: DurationInput) -> Self {
        match duration.period {
            Period::Months => Self::Months(duration.units),
            Period::Days => todo!(),
        }
    }
}

impl From<DurationInput> for DomainObligationDuration {
    fn from(duration: DurationInput) -> Self {
        match duration.period {
            Period::Months => todo!(),
            Period::Days => Self::Days(duration.units.into()),
        }
    }
}

impl From<DurationInput> for Option<DomainObligationDuration> {
    fn from(duration: DurationInput) -> Self {
        match duration.period {
            Period::Months => todo!(),
            Period::Days => Some(DomainObligationDuration::Days(duration.units.into())),
        }
    }
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize)]
pub struct CVLPctValue(rust_decimal::Decimal);
async_graphql::scalar!(CVLPctValue);

impl From<CVLPctValue> for DomainCVLPct {
    fn from(input: CVLPctValue) -> Self {
        DomainCVLPct::from(input.0)
    }
}
